src/basedtyping/__init__.py

```diff
@@ -591,6 +574,113 @@ def deco(fn: Callable[[], None]) -> Callable[[], None]: ...
         @deco
         def foo(): ...
     """
-    if not isinstance(fn, FunctionType):  # type: ignore[redundant-expr]
+    if not isinstance(fn, FunctionType):
         raise TypeError(f"{fn} is not a FunctionType")
-    return fn  # type: ignore[unreachable]
+    # https://github.com/KotlinIsland/basedmypy/issues/745
+    return cast("FunctionType[P, T]", fn)
+
+
+class ForwardRef(typing.ForwardRef, _root=True):  # type: ignore[call-arg,misc]
+    """
+    Like `typing.ForwardRef`, but lets older Python versions use newer typing features.
+    Specifically, when evaluated, this transforms `X | Y` into `typing.Union[X, Y]`
+    and `list[X]` into `typing.List[X]` etc. (for all the types made generic in PEP 585)
+    if the original syntax is not supported in the current Python version.
+    """
+
+    # older typing.ForwardRef doesn't have this
+    __slots__ = ["__forward_module__", "__forward_is_class__"]
+
+    def __init__(self, arg: str, *, is_argument=True, module: object = None, is_class=False):
+        if not isinstance(arg, str):  # type: ignore[redundant-expr]
+            raise TypeError(f"Forward reference must be a string -- got {arg!r}")
+
+        # If we do `def f(*args: *Ts)`, then we'll have `arg = '*Ts'`.
+        # Unfortunately, this isn't a valid expression on its own, so we
+        # do the unpacking manually.
+        arg_to_compile = (
+            f"({arg},)[0]"  # E.g. (*Ts,)[0] or (*tuple[int, int],)[0]
+            if arg.startswith("*")
+            else arg
+        )
+        try:
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore", category=SyntaxWarning)
+                code = compile(arg_to_compile, "<string>", "eval")
+        except SyntaxError:
+            # Callable: () -> int
+            if "->" not in arg_to_compile:
```

Comment:
would `Literal["->"]` incorrectly return `False` here?

Reply:
no, but `syntax error ->` would

Reply:
i fixed it (i think)